### ISSUE TITLE
Диаграмму визуально подчистить.

### ISSUE THREAD

**Issue by audetv** [terratensor/svodd#42]
- В графике выравнивание текста с количеством комментариев в синем поле «сводд» можно сделать по правой границе, соответственно выравнивание текста остальных комментариев сайта — по левой. Тогда они будут рядом. 

- Можно убрать линии из-под сине-серого графика, равно как и проценты внизу таблицы, это упростит внешний вид. 

- Резиновая вёрстка не очень удачно увеличивает масштаб таблицы при развёртывании браузера на фулскрин FHD или UHD, строки становятся высокими и занимают слишком много места по вертикали, также увеличивается расстояние между ними, при этом уменьшенное окно браузера показывает намного более читаемую таблицу. Опытным путём установлено, что максимальное увеличение для этого графика должно составлять не более 60% от его текущего максимума, а вообще, момент перехода между скрытым полем с адресами тем, и его появлением — вот тогда в принципе уже нормальный масштаб, его достаточно. 

- К слову, текст ссылки в левой части таблицы можно сократить или вообще удалить, заменив его на номер темы и даты её актуальности, например «12 тема с 20 мая по 12 июня 2022 года» или что-то подобное, сейчас визуально это забор адресов. 

- Текст в заголовке страницы с диаграммой можно укоротить до одной строки, а сам заголовок СВОДД вообще убрать, поскольку он дублирует и так выделенное название страницы в меню. 

- Кнопка «все темы одним потоком» слишком избыточная. Особенно это заметно в резиновой вёрcтке, если развеhнуть браузер на фулскрин на компьютере — кажется, увеличивается размер шрифта. Можно попробовать вообще от неё избавиться, объединив уже существующий пункт меню «обсуждение», который ведёт на ту же страницу. Что-то вроде «обсуждение одним потоком» или «всё обсуждение» и тд.

_Originally posted by @iprst in https://github.com/audetv/fct-search/issues/37#issuecomment-1496308648_

**Comment by audetv** [comment]
Нужно ли увеличить высоту диаграммы в браузере фулскрин? сейчас 80vh, при увеличении этого параметра увеличивается высота горизонтальных столбцов.
Сейчас все помещается в один экран, но это пока тем 32, потом же их будет больше.

**Comment by iprst** [comment]
Увеличивать не нужно, отлично умещается, хороший масштаб. Посмотрим, когда тем будет в полтора раза больше. 

Во всплывающей подсказке над графиком и из легенды можно убрать значки # — они только загромождают. Также во всплывающей подсказке нужно добавить знак процента к величине процента, вот это не очевидно читается. 

Если ничего не путаю, сейчас в графике есть небольшое расхождение процентовки и количества комментариев. После открытия новой темы, откуда номер последнего коммента забирается автоматически, в старую тему могут добавляться ещё более новые комментарии, но они будут проигнорированы в автоматическом режиме, и зачтутся в общий набор. Номера и количество новых комментариев в старой теме я заношу вручную некоторое время после переезда, обычно закрытая «жива» ещё около недели. В принципе можно проверять один раз (или реже) все темы после открытия новой, а закрытую отслеживать автоматом до открытия следующей темы.

**Comment by audetv** [comment]
1. Изменён цвет кнопки поиска
2. В функции updateList свойство window.outerHeight заменено на Screen.availHeight. Теперь высота экрана в мобильных устройствах считается корректно и диаграмма не «прыгает».
3. Во всплывающей подсказке над графиком и из легенды убраны значки #
4. Во всплывающей подсказке добавлен знак процента к величине процента

> Если ничего не путаю, сейчас в графике есть небольшое расхождение процентовки и количества комментариев. После открытия новой темы, откуда номер последнего коммента забирается автоматически, в старую тему могут добавляться ещё более новые комментарии, но они будут проигнорированы в автоматическом режиме, и зачтутся в общий набор. Номера и количество новых комментариев в старой теме я заношу вручную некоторое время после переезда, обычно закрытая «жива» ещё около недели. В принципе можно проверять один раз (или реже) все темы после открытия новой, а закрытую отслеживать автоматом до открытия следующей темы.

Да. Понятны расхождения. Я когда писал механизм, собирал данные в бд, сохранял номера последнего(закрывающего комментария) и открывающего комментария в след. теме. И в тот момент, был выбор, какой комментарий посчитать закрывающим. я выбрал тот, где указана ссылка на следующее обсуждение, понимая, что будет погрешность добавил в таблицу поле delta. которое сейчас никак не рассчитывается, т.е. мысль была, что через более продолжительный промежуток делать коррекцию, но пока не реализовал. 
start_datetime - дата открывающего комментария, если не проставлять вручную
end_datetime - дата закрывающего
Тут в комментариях к классу я описал свойства https://github.com/audetv/fct-search/blob/850b9af6ee581d5affac0d83b8605725c8cf2e7c/app/src/Svodd/Entity/Chart/Data.php

Последние 2 записи таблицы выглядят так:

|id| question_id| topic_number | title | start_datetime | end_datetime | start_comment_data_id | end_comment_data_id | comments_count | comments_delta | active |
|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|
|31|6006|31|27.03.2023 |2023-03-12 19:28:00.000|2023-03-26 17:40:00.000|432513|434105|822|0|false |
|32|      32649|          32|Текущая тема|2023-03-26 17:40:00.000|2023-04-04 11:24:00.000|               434106|             435027|           670|             0|true  |

Я в своей идее хотел учесть комментарии, которые оказались не опубликованными и вычесть их из счетчика в дельте
Но не стал учитывать, те комментарии после объявления о переходе в следующую тему обсуждения. 

> . В принципе можно проверять один раз (или реже) все темы после открытия новой, а закрытую отслеживать автоматом до открытия следующей темы.

По идее так и надо, то что я не стал усложнять себе на первом этапе, теперь можно, когда основной механизм работает.
Т.е. в итоге корректировать данные со временем, раз в минуту пересчитывать соотношение через разницу data_id первого и последнего комментария, а позже делать коррекцию отдельными процессами, запускать фоном, раз в сутки например.

**Comment by audetv** [comment]
По поводу закрытия, открытия новой темы я пока не готов, не подошел к реализации, только наметил шаги.
Но вот какие мысли есть, и я потом опубликую в общей СВОДД, сейчас сюда напишу, попробую оформить лексически.
Идея с хешем хороша тем, что просто реализовать. хеш известен, ищи его по вхождению в каждом комментарии по мере поступления. Можно так и сделать, но сегодня утром посетила мысль посмотрел в поиске и установил, что новую тему называли как:
ПРОДОЛЖЕНИЕ ОБСУЖДЕНИЯ
https://фкт-алтай.рф/qa/question/view-8820

это было использовано 11 раз, еще использовали Новая тема, продолжение и пр. 
Я хочу просмотреть все закрывающие комментарии и определить эти стоп выражения, пока не сделал это, только пробежался выборочно. По идее по ним можно закрывать тему, далее проверять по ссылке после искомой фразы или перед фразой, указанный адрес на предмет открывающего комментария, в котором, так уже сложилось, что указывают списком все предыдущие темы, которые уже известны программе. таким образом можно по этим признакам уже с высокой вероятностью сделать предположение, что эта новая тема, далее например понаблюдать за ней на предмет роста кол-во комментариев, и в итоге закрепить, что это новая тема 100% или сделать коррекцию и оповестить оператора... пока такие мысли. С хешем конечно проще)

**Comment by iprst** [comment]
> какой комментарий посчитать закрывающим. я выбрал тот, где указана ссылка на следующее обсуждение, понимая, что будет погрешность добавил в таблицу поле delta. которое сейчас никак не рассчитывается, т.е. мысль была, что через более продолжительный промежуток делать коррекцию, но пока не реализовал

Если этой дельтой считать количество комментариев в старой теме, опубликованных после закрывающего (например, добавить правило относительно end_datetime), то затем эту дельту можно вычитать из текущей статистики «остального сайта», и этого будет достаточно. 

> сегодня утром посетила мысль посмотрел в поиске и установил, что новую тему называли как:
> ПРОДОЛЖЕНИЕ ОБСУЖДЕНИЯ

Проблема с чистым языковым вводом и алгоритмом распознания события в том, что текст в комментариях могут использовать случайно, например процитировав по какой-то причине его в новой теме, или набрав тот же текст «продолжение обсуждения» и тд. Это событие ненулевой вероятности. И открыв ссылку в этом комментарии в новой теме, в этой же новой теме алгоритм увидит открывающий комментарий, и понеслась _рекурсия глупоты_. С хэшем случайностей не будет, поскольку его нужно целенаправленно посчитать под номер следующей темы, однако в этом же его слабое место — его нужно специально куда-то идти и считать, таким образом не исключены какие-то ошибки или непонимание и усложнение, а подразумевается, что функционал открытия / закрытия должен быть доступен всем пользователям, без избранности, то есть неподготовленным непрограммистам настолько же легко, как подготовленным. 

Вероятно стоит ограничиться установленной формой закрывашки (и открывашки) на нормальном языке, но специально сформированной фразой, включающей номер следующей темы, может быть «СВОДД 33» дабы избежать грамматических ошибок в числительных. То есть, если комментарий содержит текст 'ПРОДОЛЖЕНИЕ ОБСУЖДЕНИЯ СВОДД ##', номер совпадает с ожидаемым, и по ссылке имеется комментарий со строкой 'Обратная хронология обсуждения СВОДД, события по' — значит правило выполняется и новая тема поступает в обработчик. Набор фраз можно сформировать из уже использовавшихся и подобных словесных форм, главное чтобы алгоритм подхвата ожидал номер новой темы. 

А здесь возникает новый фактор. Контур предикции события «открытие новой темы» можно подключать по достижению текущей темы определённого порога количества комментариев, например 750, тогда можно будет исключить ложные срабатывания, если возникнут какие-то ошибки или злоупотребления. Достаточно будет правила:
«количество комментариев в теме больше 750?» 
«нет»
«давай дасвидания»
«да»
«по ссылке есть текст открывашка?» 
и тд

**Comment by iprst** [comment]
Я пока добавлю в ближайшую закрывашку шаблон —

'ПРОДОЛЖЕНИЕ ОБСУЖДЕНИЯ СВОДД ##'

— чтобы к нему привыкали, а затем откомменчу этот момент в закрывающейся теме, чтобы была понятна суть, и дам ссылку на эту страницу с обсуждением, и можно добавить специальную страницу «шаблон открытия новой темы и закрытия старой» типа как _readme_.

**Comment by audetv** [comment]
> Если этой дельтой считать количество комментариев в старой теме, опубликованных после закрывающего (например, добавить правило относительно end_datetime), то затем эту дельту можно вычитать из текущей статистики «остального сайта», и этого будет достаточно.

Да, хорошо, согласен, реализую механизм и скорректирую статистику.

> Проблема с чистым языковым вводом и алгоритмом распознания события в том, что текст в комментариях могут использовать случайно, например процитировав по какой-то причине его в новой теме, или набрав тот же текст «продолжение обсуждения» и тд. 

Все верно и это причина моих размышлений, с одной стороны хочется автоматизировать процесс, а с другой стороны не ограничивать произвол. 

> Я пока добавлю в ближайшую закрывашку шаблон —
> 
> 'ПРОДОЛЖЕНИЕ ОБСУЖДЕНИЯ СВОДД ##'

Да, давайте остановимся на этом варианте, вы уже так сделали вчера с публикацией  темы 33.
Я начал делать механизм, доделаю команду для перехода на новую тему, вчера она у меня сбоила, плюс появились недочеты в графике - исправил, и недочеты с обновлением данных после перехода - это еще в работе.
Созданную команду можно запускать по условию, которое проверяет другой алгоритм:

> специально сформированной фразой, включающей номер следующей темы, может быть «СВОДД 33»

И этот алгоритм можно тестировать, менять и т.д. в процессе обсуждения, не затрагивая механизм смены темы.
Это планирую завершить к следующей теме и уже опробовать в рабочих условиях.

**Comment by iprst** [comment]
> Контур предикции события «открытие новой темы» можно подключать по достижению текущей темы определённого порога количества комментариев, например 750,

Подумал ещё, в данный момент это скорее глупая идея нежели полезная. Трата времени и ненужное ограничение, которое может быть полезным если возникнут какие-то злоупотребления, которых пока что ни разу не возникало.

**Comment by audetv** [comment]
Это да, злоупотреблений пока не возникало, я пока сосредоточусь на основном функционале, эта идея в копилку.

Программе - новому сервису надо отследить условную фразу и каким то образом передать эту информацию сервису parser, парсер я делал, как независимую консольную утилиту, и он не имеет доступа к базе данных, все данные берет с сайта ФКТ и через ввод параметров в консоли. В данном случае ему надо давать на вход новый url, который он будет успешно разбирать. 

Сейчас решаю архитектурную задачу, как это сделать, пока представляется, что должен быть некий новый сервис, не знаю, типа индексёр служба. которая будет принимать на вход команды и запускать нужные части системы(сервисы), т.е., например парсер запускать с новыми параметрами и урл.  Или можно, конечно, наделить парсер дополнительной функциональностью, пока не решил. И это походе приводит к очередям сообщений.

Очереди подразумевают переход на другую модель запуска задач в приложении, т.е когда происходит некоторое событие и достигается условие, создается event или сообщение и передается в очередь на обработку, подписчики (службы) получают сообщения в свой адрес и запускают нужные команды. В общем то все, что я выше написал, это стандартное описание своими словами шаблона разработки event -listener -subscriber событие, подписка, очереди.
Просто надо сделать нашу конкретную реализацию, но пока не просто, пока в процессе, формирую нужные связи)

**Comment by iprst** [comment]
Сервис слежения это модуль, который перебирает новые id в базе комментариев с целью проставить отметку в поле «является ли этот комментарий закрывающим»:
— коммент номер абвгд содержит форму закрывашку?
— ДА (записываем в базу, отправляем событие модулю обработки команд) / НЕТ (записываем в базу, проверяем следующий номер id).

Если говорить о дополнительном функционале, то нужно исходить из потребностей. Сделать то можно что угодно, но всё требует ресурсов, поэтому делать нужно только то что нужно, а что не нужно, делать не нужно (с) Винни Пух. То есть если есть какие-то идеи, куда применить поля из базы и что нарисовать на экране, то под это и нужно допиливать парсер. Базу ведь можно в любой момент снабдить произвольным количеством новых полей и собрать дополнительные данные, когда потребуется. 

Если нужны идеи, что можно сделать за графики, то у меня есть, я сам пока не дорос до разработки, но некие полезные мысли были. Это наверное куда-то в раздел проектов или предложений надо отправлять, чтобы не захламлять вопрос графика / диаграммы?

Ещё по диаграмме момент всплыл: сейчас наверху страницы идёт график, а ниже идёт список тем, но ведь в графике уже есть список тем, почему бы не сделать внутри него ссылки, чтобы не дублировать по сути ту же информацию на странице?

**Comment by iprst** [comment]
Странное поведение списка тем в диаграмме, которое ранее не наблюдалось: при некоторых размерах окна и при определённом масштабе, из диаграммы исчезают все чётные строки текста, график остаётся нетронутым. Поведение наблюдается на разных масштабах и при разном размере окна, и зависит от соотношения этих параметров, например масштаб 90% на некоторых размерах окна даёт эффект, а на других размерах окна всё нормально. 

![chrome_gnugz5ZyIO](https://github.com/terratensor/svodd/assets/129882753/3b4b324c-3e0d-449a-a1e3-9c361d0d972c)

**Comment by audetv** [comment]
Да, хорошо. это известная мне проблема, я с ней столкнулся с самого начала, еще когда сделал диаграмму, периодически решал вопрос внесением изменений в код js , меняя высоту окна, по мере добавления новых тем. Последнее время этого не делал. надо будет обновить и еще раз подумать, должно быть решение, надо будет подумать и придумать, которое освободит от обновления кода по мере добавления тем. 

Если визуально то проблем здесь, надо как-то заранее рассчитать размер окна и высоту, когда подключал библиотеку для графика, такого функционала не нашел, видимо это не особо востребованный функционал) А возможно в тот момент не смог его распознать, нужно еще раз выделить время для изучения.

<details><summary>Details</summary>
<p>

![2024-04-26_11-39-56](https://github.com/terratensor/svodd/assets/10896447/c57bdbe0-5ed8-44c8-8101-d6f34fa7b770)

![2024-04-26_11-40-18](https://github.com/terratensor/svodd/assets/10896447/59729275-ccfa-4563-91d5-dc70475ecf6b)

</p>
</details> 

Вопрос ранее вы делились названием программ, с помощью которой можно легко делать анимированные скриншоты, типа как вы сделали сейчас. Я тогда не сохранил себе информацию и не успел воспользоваться, а позже потерял. Поиском на гитхабе у меня не получается нормально пользоваться, он как-будто бы не для поиска, а для чего то другого сделан. Очень сложно найти то, что обсуждалось ранее, поделитесь наименованием программы снова)

**Comment by iprst** [comment]
> делились названием программ, с помощью которой можно легко делать анимированные скриншоты

ShareX
https://getsharex.com
